Support escapes in attributes

kmaasrud opened the pull request:
This replaces the `CowStr` in `Attributes` with a new struct: `Attr`. This struct implements `Display`, which skips over what needs to be skipped and writes everything else.

~~If we want to keep the raw attributes as string slices, we cannot continue storing multiple classes as a space-separated string, but rather as a vec of string slices. This is also better when it comes to other output formats, as space-separated classes could be considered an HTML-ism. Seeing as classes really are a different type of attribute than all others, I am a bit unsure how we should do this. I toyed around with making `Attr` an enum (variants being `Class(Vec<&str>)` and `Other(&str)`,) but the nesting made this a nightmare. @hellux I would love your input on this!~~

~~Additionally, I have not yet made the `escape_backslash` test pass yet, but I have some ideas on how to get that going. Shouldn't be that difficult.~~

hellux reviewed (commented):
Quick benchmarks are easy to do and should be done early, as it impacts the implementation and API. E.g. the `cargo bench -p bench-iai` benchmark is pretty good at giving quick feedback whether performance has improved or worsened.

Regarding the storage of attributes, I have been experimenting on a branch to get rid of the DiscontinuousStrings object (to avoid needing whole block beforehand) and increase performance, which led to some experimental changes in the attr module: https://github.com/hellux/jotdown/blob/inline_perf/src/attr.rs. The spans are no longer guaranteed to be continuous for the inline module, so it must provide multiple discontinuous spans in case they are on multiple lines. So for attributes I tried using a vec with a sequence of keys/values, where the values might be one or more if needed. Currently each set of attributes is a vec but it might be useful to use a single shared vec for all attributes, to reduce allocations. Also, URLs and link labels may be stored here also, as they may also span multiple lines.

So the Attributes::get method returns a ValueRef object which has a parts method which returns an iterator of str's. Usually it's a single str but if the value spans multiple lines or if it has multiple class values it will be several str's. E.g. `.a .b` yields ["a", "b"] for get("class"). To combine this with your escaped attributes it could perhaps return an iterator of chars instead.

Does this seem like a reasonable implementation? It seems slightly complicated but not sure if it is possible to make it much simpler without just allocating strings like currently.

Regarding space-separated classes being HTML-specific, the reference implementation AST does not seem to distinguish between ".a .b" and "class="a b". Not sure if we need to either, the whole class and id things are already inherited from HTML.

kmaasrud commented on src/attr.rs:
```diff
@@ -88,12 +119,12 @@ impl<'s> Attributes<'s> {
             let prev = &mut attrs[i].1;
             if key == "class" {
-                *prev = format!("{} {}", prev, val).into();
+                // prev.raw = format!("{} {}", prev, val);
```
Exactly! Not sure if there really is a way to do this without allocating a new string (by making `Attr::raw` a CowStr). That's partly why I started thinking about storing classes as a Vec of string slices, but I agree that it makes sense to keep the HTML-ism

hellux replied:
Yeah, might be best to use cowstr for now. If we later go with the method on the `inline_perf` branch (which is kind of a vec of string slices) we will get rid of them completely.

kmaasrud replied:
Using CowStr now, and the tests pass!

kmaasrud commented on src/attr.rs:
```diff
@@ -44,17 +73,19 @@ impl<'s> Attributes<'s> {
+                Cow::Owned(_) => panic!(),
+                Cow::Borrowed(s) => s,
+            }
+        }
```
Yes agreed! Should probably open a new issue for it

kmaasrud replied:
Opened: #20

kmaasrud commented on src/attr.rs:
```diff
@@ -394,6 +426,23 @@ mod test {
+        );
+        test_attr!(
+            r#"{key="should not be a \% comment"}"#,
+            ("key", "should not be a % comment")
```
No, I quickly realised that too, just kept it for testing.

kmaasrud commented on src/attr.rs:
```diff
@@ -394,6 +426,23 @@ mod test {
+            r#"{key="should not be a \% comment"}"#,
+            ("key", "should not be a % comment")
+        );
+    }
```
Indeed

kmaasrud commented on src/html.rs:
```diff
@@ -194,7 +194,7 @@ impl<'s, I: Iterator<Item = Event<'s>>, W: std::fmt::Write> Writer<'s, I, W> {
                     for (a, v) in attrs.iter().filter(|(a, _)| *a != "class") {
                         write!(self.out, r#" {}=""#, a)?;
-                        self.write_escape(v)?;
+                        self.write_escape(&format!("{}", v))?;
```
As we already encourage buffering on sinks sensitive to many small writes, I guess we can just use the `Attr::value` method and write each char (respecting escapes), like you propose?

kmaasrud replied:
Fixed here:

https://github.com/hellux/jotdown/blob/7529059f47d1ab174cd1f4493f52dca68f9c61a5/src/html.rs#L197

hellux replied:
did a quick benchmark by trying to use the write_escaped_char in write_escape:
```diff
-    fn write_escape(&mut self, mut s: &str) -> std::fmt::Result {
-        let mut ent = "";
-        while let Some(i) = s.find(|c| {
-            match c {
-                '<' => Some("&lt;"),
-                '>' => Some("&gt;"),
-                '&' => Some("&amp;"),
-                _ => None,
-            }
-            .map_or(false, |s| {
-                ent = s;
-                true
-            })
-        }) {
-            self.out.write_str(&s[..i])?;
-            self.out.write_str(ent)?;
-            s = &s[i + 1..];
-        }
-        self.out.write_str(s)
+    fn write_escape(&mut self, s: &str) -> std::fmt::Result {
+        s.chars().try_for_each(|c| self.write_escaped_char(c))
     }
```

It seems like it is considerable slower to go char by char than trying to write strs (even when writing to buffered output):
```
  html/readme             time:   [39.416 µs 39.464 µs 39.540 µs]
                        thrpt:  [22.838 Melem/s 22.882 Melem/s 22.910 Melem/s]
                 change:
                        time:   [+76.224% +93.766% +127.16%] (p = 0.00 < 0.05)
                        thrpt:  [-55.979% -48.391% -43.254%]
```

However, it might be difficult to avoid to go char by char for the escapes (unless we expose the raw str?). Currently, the HTML output is around 10% slower on this branch compared to master, but the output is by far the fastest part so it is does not have a considerable impact, I think.

kmaasrud replied:
We could make an iterator that iterates over string slices of `Attr::raw` instead? That seems similar to what you have done in the `inline_perf` branch. But as long as we only use `write_escaped_char` for writing the attributes, then we shouldn't be in such a bad shape, I agree.

hellux replied:
> We could make an iterator that iterates over string slices of `Attr::raw` instead? That seems similar to what you have done in the `inline_perf` branch.

yeah, could be worth a try

kmaasrud replied:
Done!

kmaasrud commented on src/lib.rs:
```diff
@@ -547,7 +547,7 @@ impl<'s> PrePass<'s> {
                         .as_ref()
                         .and_then(|attrs| attrs.get("id"))
-                        .map(ToString::to_string);
+                        .map(|id| format!("{}", id));
```
I don't think this is possible at all in this case... `id_override` is inserted into the publicly exposed `Container::Heading::id` directly, which means the only other solution would be to expose it to the user as an iterator of chars. I don't think that would be very ergonomic. Thoughts?

hellux replied:
Yeah, true, probably better to keep it like this for now. A question for later might be if we want to try to create some ergonomic type for these possibly non-continuous strings, e.g. URLs, link tags, heading ids.

kmaasrud replied:
Agreed!

hellux commented on src/attr.rs:
```diff
@@ -105,15 +159,13 @@ impl<'s> Attributes<'s> {
-    pub fn get(&self, key: &str) -> Option<&str> {
+    pub fn get(&self, key: &str) -> Option<&Attr<'s>> {
         self.iter().find(|(k, _)| *k == key).map(|(_, v)| v)
     }
```
because `Attr` may only contain a single continuous `&str`, it is impossible to represent concatenated classes. `attrs.get("class")` here will only return the first class, it seems.

hellux replied:
For now, maybe we should use an owned cowstr in those cases?

kmaasrud replied:
Yes, that's exactly what I was thinking in this https://github.com/hellux/jotdown/pull/19#discussion_r1127039234. I agree, a CowStr seems reasonable for the time being.

kmaasrud replied:
Using CowStr now.

hellux commented on src/attr.rs:
```diff
@@ -24,12 +29,59 @@ pub fn valid<I: Iterator<Item = char>>(chars: I) -> (usize, bool) {
+                        return Some('\\');
+                    }
+                }
+                _ => return Some(c),
```
do we want to match the reference implementation by only  escaping ascii punctuation? E.g.

- `\a` -> `\a`
- `\.` -> `.`
- `\\` -> `\`

The lexer behaves like this also, so it might be more "unified" to match this behavior.

kmaasrud replied:
Done!

hellux commented on src/attr.rs:
```diff
@@ -231,13 +290,19 @@ impl<I: Iterator<Item = char>> Parser<I> {
-                    } else {
+                ValueQuoted => match c {
+                    '\\' => {
+                        if let Some(c) = self.chars.next() {
```
I think it would be preferable to add a new state here to the FSM instead of adding another place where we consume a char. Easier to keep track of a state machine where each step is a single char.

hellux replied:
Or actually, it might make things more complicated due to how the next/step functions work right now. Maybe best to keep it like this for now.

hellux commented on src/html.rs:
```diff
@@ -450,4 +450,13 @@ impl<'s, I: Iterator<Item = Event<'s>>, W: std::fmt::Write> Writer<'s, I, W> {
+            '<' => self.out.write_str("&lt;"),
+            '>' => self.out.write_str("&gt;"),
+            '&' => self.out.write_str("&amp;"),
+            c => self.out.write_char(c),
```
We should actually also escape `"` here as `&quot;`. I didn't think it was needed earlier because it always turns into either a left or right quote in formatted text.

There is actually a test from the reference implementation for this that we should enable now:

```
---- suite::attributes::test_48546bb stdout ----
thread 'suite::attributes::test_48546bb' panicked at 'assertion failed: `(left == right)`
  left: `"<p><span key=\"\"#hi\">hi</span></p>"`,
 right: `"<p><span key=\"&quot;#hi\">hi</span></p>"`:
========================= INPUT ============================
hi{key="\"#hi"}
=================== ACTUAL vs EXPECTED =====================
<p><span key=""#hi">hi</span></p>    !=    <p><span key="&quot;#hi">hi</span></p>
============================================================
', tests/suite/attributes.rs:83:5
```

Actually, it should be escaped in the write_escape function also, currently we can get invalid HTML if we use `"` in urls, for example:
```
[text]("url)]
```
```html
<p><a href=""url"">text</a></p>
```

hellux replied:
Hmm, I was thinking so at first. But it should be possible to have unescaped quotes in normal text, as shown by the failing unit test.

I suppose we must have a flag of some sort, then. For normal text it should be unescaped, within attribute strings it should be escaped. We could have an `is_attribute` flag, and if it is set, we escape the quote. However, a bool is quite unreadable from the call side.

How about if we add a flag to `write_escape` and then we add two functions `write_text` and `write_attr` that simply calls the `write_escape`? And make sure the write_escape is never used directly.

kmaasrud replied:
Fixed!

hellux commented on src/attr.rs:
```diff
@@ -24,12 +26,55 @@ pub fn valid<I: Iterator<Item = char>>(chars: I) -> (usize, bool) {
+}
+
+impl<'s> Attr<'s> {
+    pub fn write<F: FnMut(&str) -> fmt::Result>(&self, mut write_fn: F) -> fmt::Result {
```
Is there an advantage of this method over returning a &str iterator? A &str iterator seems more general? Which is useful when it is part of the public API.

We should be able to do simply
```rs
attr.parts().try_for_each(|p| self.write_escape(p))?;
```
in html.rs then.

kmaasrud replied:
The limitation mainly comes from needing more mutable state than the AttrChars iterator. This comes from e.g. the escape state toggle, no backslashes added to the first part, etc. I could of course add these to an iterator struct, but that would add more code bloat, and I simply thought this way would be simpler.

If you prefer iterators, then I have no issue with changing to it :smile:

hellux replied:
Yeah, in my opinion, iterators seem more convenient from the client pov, otherwise you are tied to the str -> format::error function signature. You might not always be interested in writing, could be reading only for example. I think the performance should be more or less the same between the two different methods.

I guess we are not forced to have a dedicated iterator struct, could just use std::iter::from_fn and return an impl Iterator<&str>. But a struct may be more user-friendly because it can be named in type signatures, I suppose.

kmaasrud replied:
Got it! Will add ASAP

kmaasrud replied:
Implemented now, with the `AttributeValueParts` iterator.

hellux commented on src/attr.rs:
```diff
@@ -24,12 +26,55 @@ pub fn valid<I: Iterator<Item = char>>(chars: I) -> (usize, bool) {
+}
+
+impl<'s> Attr<'s> {
+    pub fn write<F: FnMut(&str) -> fmt::Result>(&self, mut write_fn: F) -> fmt::Result {
```
Since this is part of the public API, there should probably be a doc comment explaining what it does exactly and what the purpose of the function is, i.e. split on escaped chars in order to avoid string allocation of discontinous strings.

kmaasrud replied:
`AttributeValue`, `AttributeValueParts` and their public API is now documented.

kmaasrud commented:
All comments are adressed, and the whole suite now passes! I renamed the structs to `AttributeValue` and `AttributeValueParts` to make them a bit more clear, and I also added them to the public API with some simple docstrings.

hellux commented on src/attr.rs:
```diff
@@ -24,12 +27,90 @@ pub fn valid<I: Iterator<Item = char>>(chars: I) -> (usize, bool) {
+pub struct AttributeValueParts<'s> {
+    split: Split<'s, char>,
+    escape: bool,
+    next: Option<&'s str>,
```
Two thoughts about this iterator:

- the size seems quite large, it is currently 96 bytes, not sure how the Split
  looks on the inside but it seems surprisingly large.
- currently it seems to return e.g. `"a" + "\" + "b"` for `"a\\b"`, but ideally
  it should return `"a" + "\b"`, right?

I'm thinking it should be possible with a single &'str (16 bytes). E.g for the
string `"a\.b\c\\."` we get three strs emitted:

- new: `ahead <= "a\.b\c\\."`,
- next: first backslash has punctuation after
    => `ahead <= ".b\c"`, return `Some("a")`
- next: second backslash has punctuation after,
    => `ahead <= "\."`, return `Some(".b\c")`
- next: backslash at pos 0, so we assume it has already been escaped
    => `ahead <= ""`, return `Some("\.")`
- next: ahead is empty
    => return None

In the case where it starts with a backslash:

- `"\abc"` => new: `ahead <= "\abc"`, so backslash gets emitted,
- `"\.abc"` => new: `ahead <= ".abc"`, so the backslash does not get emitted,

It's not necessary to try this out right now, though.

kmaasrud replied:
I have to say; you are a wizard! Not only did that method reduce the memory footprint of the iterator, I also found a way to split the string into as few as possible splits while implementing it. All tests still pass, and this is definitely the way to go!

hellux commented on src/attr.rs:
```diff
@@ -24,12 +27,90 @@ pub fn valid<I: Iterator<Item = char>>(chars: I) -> (usize, bool) {
+/// Stores an attribute value that supports backslash escapes of ASCII punctuation upon displaying,
+/// without allocating.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct AttributeValue<'s> {
```
since the module is called `attr` we could perhaps call this `attr::Value` instead of `attr::AttributeValue`? same with `attr::ValueParts` in that case.

kmaasrud replied:
Reasonable, indeed! Although then perhaps we should not expose the structs of the `attr` module directly to the library root? Having `Value` and `ValueParts` in the top-level of the docs could be a bit confusing (same applies for `use` statements.)

hellux replied:
Oh, yes, true, I forgot to consider the public API. I don't think we want to expose the `attr` module, as it seems like an implementation detail. Simply exposing `Attributes`, `AttributeValue` and `AttributeValueParts` is probably best. We could either keep the full name `AttributeValue` or do `pub use attr::Value as AttributeValue`. I think the the first option might be best, because then it is clear that only one type exists.

kmaasrud replied:
Done!

hellux reviewed (commented):
It looks good now, I think it is ready to merge! I just gave some thoughts on your last changes, feel free to address them however you want.

kmaasrud commented on src/lib.rs:
```diff
@@ -547,7 +547,7 @@ impl<'s> PrePass<'s> {
                         .as_ref()
                         .and_then(|attrs| attrs.get("id"))
-                        .map(ToString::to_string);
+                        .map(|id| format!("{}", id));
```
You are right, I totally forgot the blanket impl of `ToString`. Change removed.

hellux commented:
one last thing, there are two tests (48546bb, 6bc4257) from the reference implementation that previously failed but are passing after this commit. They can be removed from `tests/suite/skip` so that they are run from now on.

hellux commented on src/html.rs:
```diff
@@ -249,7 +249,7 @@ impl<'s, I: Iterator<Item = Event<'s>>, W: std::fmt::Write> Writer<'s, I, W> {
                             }
                             first_written = true;
-                            self.out.write_str(cls)?;
+                            write!(self.out, "{}", cls)?;
```
this should actually use write_attr, the write_str was a bug.

kmaasrud commented:
All comments fixed and I've rebased from master.

hellux commented:
Looks great, thanks for the patch!
